### PR TITLE
Intelfix

### PR DESCRIPTION
**Pull request by sbird**
This makes a whole bunch of shifts unsigned. Apologies for the whitespace changes.

### CODE REVIEW

File: libgadget/domain.c [hunk 1]
```diff
@@ -317,8 +317,8 @@ static int64_t
 domain_particle_costfactor(int i)
 {
     /* We round off GravCost to integer*/
-    if(P[i].TimeBin)
-        return (1 + P[i].GravCost) * (TIMEBASE / (1 << P[i].TimeBin));
+    if(P[i].TimeBin > 0)
+        return (1 + P[i].GravCost) * (TIMEBASE / (1u << P[i].TimeBin));
```
Comment (rainwoodman) [inline libgadget/domain.c]: TimeBin also need to be cast unsigned, unfortunately...

**Comment by rainwoodman** [comment]
Perhaps define a SHL and SHR macro that always cast both operators?

File: libgadget/forcetree.c [hunk 2]
```diff
@@ -515,7 +515,7 @@ force_tree_build_single(const struct TreeBuilder tb, const int npart)
  */
 void force_create_node_for_topnode(int no, int topnode, unsigned int bits, unsigned int x, unsigned int y, unsigned int z, int *nextfree, const int lastnode)
 {
-    unsigned int i, j, k;
+    int i, j, k;
```
Comment (rainwoodman) [inline libgadget/forcetree.c]: Why is it important to keep these signed?

File: libgadget/domain.c (hunk 1 above)
Comment (sbird) [reply libgadget/domain.c]: TimeBin has been made unsigned in the definition in partmanager.h

File: libgadget/forcetree.c (hunk 2 above)
Comment (sbird) [reply libgadget/forcetree.c]: Look at line 537: 2*i - 1, which is -1 if i == 0. if i is unsigned that comes out as 2**32 - 1. I am worried about this pull for that reason: there is the potential to introduce subtle bugs worse than the compiler bug we are working around. Please review carefully.

File: libgadget/timebinmgr.h [hunk 3]
```diff
@@ -10,7 +10,7 @@
  *   Note that because each snapshot uses TIMEBASE on the integer timeline, the conversion
  *   factor between loga and ti is not constant across snapshots.
  */
-#define TIMEBINS 20
+#define TIMEBINS 20u
 #define TIMEBASE (1u<<TIMEBINS)
 #define MAXSNAPSHOTS (1u<<(30-TIMEBINS))
```
Comment (rainwoodman) [inline libgadget/timebinmgr.h]: 30u?

Comment (sbird) [reply libgadget/timebinmgr.h]: Yes! Thank you.

**Comment by sbird** [comment]
I am wary of the macro because I fear that it could create subtle bugs. I am already a little unsettled by how large this one is, for an easily worked around bug.

**Comment by sbird** [comment]
I would appreciate your opinion on whether we should commit a more minimal version of this, actually. The bug seems quite specific and definitely a compiler bug.

**Comment by rainwoodman** [comment]
I definitely felt uncomfortable about the wide spread of 'u' everywhere. (hence the suggestion of using a macro).

Perhaps we can limit this to only the line where the bug is triggered, and add a reference to the ticket on github in the source code?

**Comment by sbird** [comment]
That sounds like a good idea. I will strip this one down.

**Comment by sbird** [comment]
Do you have a link for the upstream icc bug report?

**Comment by rainwoodman** [comment]
No. I don't think we'll have anything, but we can have an issue here, and probably update it when @marcel-sl hear anything back from the site admin.

**Comment by sbird** [comment]
I like that a lot more. Few opportunities for bugs. Let us wait until travis goes green and then commit

**Comment by rainwoodman** [comment]
Yes. It looks good.

**Pull request merged**